### PR TITLE
added logger manager and global exception handler

### PR DESCRIPTION
**Pull request by stetskivostap**

### CODE REVIEW

**Comment by vladpavlyuk** [comment]
Everything is fine, but I'd like you to not use third-party library for logging exception. Could you please write custom log writer?

File: VirtualClassRoom/Extensions/ExceptionMiddleware.cs [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+        private Task HandleExceptionAsync(HttpContext context, Exception exception)
+        {
+            context.Response.ContentType = "application/json";
+            context.Response.StatusCode = (int)HttpStatusCode.InternalServerError;
+
+            return context.Response.WriteAsync(new ErrorDetails()
+            {
+                StatusCode = context.Response.StatusCode,
+                Message = "Internal Server Error."
```
Comment (vladpavlyuk) [inline VirtualClassRoom/Extensions/ExceptionMiddleware.cs:42]: Move this string to resources

File: VirtualClassRoom/Extensions/ExceptionMiddleware.cs [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+        public async Task InvokeAsync(HttpContext httpContext)
+        {
+            try
+            {
+                await _next(httpContext);
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError($"Something went wrong: {ex}");
```
Comment (vladpavlyuk) [inline VirtualClassRoom/Extensions/ExceptionMiddleware.cs:29]: Here  the same

File: Common/Constants/FileConstants.cs [hunk 3]
```diff
@@ -0,0 +1,11 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Common.Constants
+{
+    public class FileConstants
+    {       
+        public const string FilePath = @"C:\Users\stets\OneDrive\Рабочий стол\leobitr\Virtual-Classroom\Common\LogFiles\";
```
Comment (vladpavlyuk) [inline Common/Constants/FileConstants.cs:9]: The path should be relative. It means that it should work on my computer as well, but it won’t. What’s going to happen if you move project to another folder? In your case, error logs will keep writing to `C:\Users\stets\OneDrive\Рабочий стол\leobitr\Virtual-Classroom\Common\LogFiles\";`. Please fix this

File: BLL/Services/Log.cs [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+using System;
+using System.IO;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace BLL.Services
+{
+    public class Log : ILog
+    {
+        public Log()
```
Comment (vladpavlyuk) [inline BLL/Services/Log.cs:12]: Remove this constructor. You never use it

File: VirtualClassRoom/Extensions/ExceptionMiddleware.cs [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+        public async Task InvokeAsync(HttpContext httpContext)
+        {
+            try
+            {
+                await _next(httpContext);
+            }
+            catch (Exception ex)
+            {
+                await _logger.LogException($"{ErrorConstants.Oops}: {ex}");
```
Comment (vladpavlyuk) [inline VirtualClassRoom/Extensions/ExceptionMiddleware.cs:31]: I don’t mind to add oops to error message. Since you add this text to all errors, it makes sense to add this text inside of LogException method

File: BLL/Services/Log.cs [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+        {
+        }
+
+        private static readonly Lazy<Log> instance = new Lazy<Log>(() => new Log());
```
Comment (vladpavlyuk) [inline BLL/Services/Log.cs:16]: Could you explain me why do you use `Lazy<>` here?

File: BLL/Services/Log.cs [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+        {
+            string fileName = string.Format("{0}_{1}.txt", "Exception", DateTime.Now.ToString("dd_MM_yyyy"));
+            string logFilePath = string.Format(@"{0}{1}", FileConstants.FilePath, fileName);          
+            StringBuilder sb = new StringBuilder();
+            sb.AppendLine("````````````````````````````````````````````");
+            sb.AppendLine(ErrorConstants.InternalServerError);
+            sb.AppendLine(DateTime.Now.ToString());
+            sb.AppendLine(message);
+
+            await using (StreamWriter writer = new StreamWriter(logFilePath, true))
```
Comment (vladpavlyuk) [inline BLL/Services/Log.cs:36]: Could you please use `System.IO.File.WriteAllText` instead of using the stream? This method is more flexible and will look better. Also you don’t have to dispose stream every time

File: BLL/Services/Log.cs [hunk 8]
```diff
@@ -9,35 +10,28 @@ namespace BLL.Services
...
+            string logFilePath = string.Format(@"{0}{1}{2}",rootPath,@"\Common\LogFiles\",fileName);          
             StringBuilder sb = new StringBuilder();
-            sb.AppendLine("````````````````````````````````````````````");
+            sb.AppendLine("``````````````````````````````````````````````````````````````````````````````````````````");
             sb.AppendLine(ErrorConstants.InternalServerError);
             sb.AppendLine(DateTime.Now.ToString());
             sb.AppendLine(message);
-
-            await using (StreamWriter writer = new StreamWriter(logFilePath, true))
+            if(!File.Exists(logFilePath))
```
Comment (vladpavlyuk) [inline BLL/Services/Log.cs:30]: Check if WriteAllText method creates a file, if so then remove this check on file exist

File: VirtualClassRoom/Extensions/ExceptionMiddleware.cs [hunk 9]
```diff
@@ -28,7 +28,7 @@ public async Task InvokeAsync(HttpContext httpContext)
             }
             catch (Exception ex)
             {
-                await _logger.LogException($"{ErrorConstants.Oops}: {ex}");
+                await _logger.LogException($"{ErrorConstants.InternalServerError}: {ex}");
```
Comment (vladpavlyuk) [inline VirtualClassRoom/Extensions/ExceptionMiddleware.cs:31]: Remove this string interpolation and move it to the LogException method

File: BLL/Services/Log.cs [hunk 10]
```diff
@@ -9,35 +10,28 @@ namespace BLL.Services
...
-                return instance.Value;
-            }
-        }
+            _web = web;
+        }       
 
         public async Task LogException(string message)
         {
+            int removeLength = _web.ContentRootPath.Length - 17;
+            string rootPath = _web.ContentRootPath.Remove(removeLength);
```
Comment (vladpavlyuk) [inline BLL/Services/Log.cs:22]: Remove this code

File: BLL/DTO/ErrorDetails.cs [hunk 11]
```diff
@@ -0,0 +1,16 @@
+﻿using Newtonsoft.Json;
+
+namespace BLL.DTO
+{
+    public class ErrorDetails
```
Comment (vladpavlyuk) [inline BLL/DTO/ErrorDetails.cs:5]: Delete this if you dont use it

File: BLL/Services/Log.cs [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+        {
+            _web = web;
+        }       
+
+        public async Task LogException(string message)
+        {
+            int removeLength = _web.ContentRootPath.Length - 17;
+            string rootPath = _web.ContentRootPath.Remove(removeLength);
+            string fileName = string.Format("{0}_{1}.txt", "Exception", DateTime.Now.ToString("dd_MM_yyyy"));
+            string logFilePath = string.Format(@"{0}{1}{2}",rootPath,@"\Common\LogFiles\",fileName);
```
Comment (vladpavlyuk) [inline BLL/Services/Log.cs:24]: Use this https://docs.microsoft.com/ru-ru/dotnet/api/system.io.path.combine?view=netframework-4.8

Comment (vladpavlyuk) [inline BLL/Services/Log.cs:24]: Move all strings to resources or constants

File: BLL/Services/Log.cs [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+        private readonly IWebHostEnvironment _web;
+        public Log(IWebHostEnvironment web)
+        {
+            _web = web;
+        }       
+
+        public async Task LogException(string message)
+        {
+            int removeLength = _web.ContentRootPath.Length - 17;
+            string rootPath = _web.ContentRootPath.Remove(removeLength);
```
Comment (vladpavlyuk) [inline BLL/Services/Log.cs:22]: Please, check this https://docs.microsoft.com/ru-ru/dotnet/api/system.environment.currentdirectory?view=netframework-4.8

File: BLL/DTO/ErrorDetails.cs (hunk 11 above)
Comment (stetskivostap) [reply BLL/DTO/ErrorDetails.cs:5]: Actually I use this to send response on front end, it shows error message and status code if some errors occure

File: Common/Constants/FileConstants.cs [hunk 14]
```diff
@@ -0,0 +1,18 @@
+﻿using System;
+
+namespace Common.Constants
```
Comment (vladpavlyuk) [inline Common/Constants/FileConstants.cs:3]: Delete this class because you never use it

File: BLL/DTO/ErrorDetails.cs (hunk 11 above)
Comment (vladpavlyuk) [reply BLL/DTO/ErrorDetails.cs:5]: It's not a good approach. We will implement different, but a little bit similar approach

File: BLL/Services/Log.cs [hunk 15]
```diff
@@ -15,23 +15,24 @@ public Log(IWebHostEnvironment web)
...
             sb.AppendLine(ErrorConstants.InternalServerError);
             sb.AppendLine(DateTime.Now.ToString());
-            sb.AppendLine(message);
-            if(!File.Exists(logFilePath))
-            {
-                using (File.Create(logFilePath)) { }                                                  
-            }
-            await File.WriteAllTextAsync(logFilePath, sb.ToString());
+            sb.AppendLine(message);           
+            await File.AppendAllTextAsync(path, sb.ToString());
```
Comment (vladpavlyuk) [inline BLL/Services/Log.cs:34]: Why did you change **WriteAllText** to **AppendAllText**?

File: BLL/Services/Log.cs [hunk 16]
```diff
@@ -15,23 +15,24 @@ public Log(IWebHostEnvironment web)
         {
             _web = web;
         }       
+        
+        private string GetFilePath()
+        {
+            var fileName = string.Format("{0}_{1}.txt", "Exception", DateTime.Now.ToString("dd_MM_yyyy"));
+            var logFilePath = string.Format(@"{0}{1}{2}", _web.ContentRootPath, @"\LogFiles\", fileName);
```
Comment (vladpavlyuk) [inline BLL/Services/Log.cs:22]: I asked you to not use hardcoded strings. Please move them to constants

File: BLL/Services/Log.cs [hunk 17]
```diff
@@ -15,23 +15,24 @@ public Log(IWebHostEnvironment web)
         {
             _web = web;
         }       
+        
+        private string GetFilePath()
+        {
+            var fileName = string.Format("{0}_{1}.txt", "Exception", DateTime.Now.ToString("dd_MM_yyyy"));
```
Comment (vladpavlyuk) [inline BLL/Services/Log.cs:21]:
```suggestion
            var fileName = string.Format(ExceptionConstants.FileNamePattern, DateTime.Now.ToString(ExceptionConstants.DateTimeFormat));
```

File: BLL/Services/Log.cs (hunk 16 above)
Comment (vladpavlyuk) [inline BLL/Services/Log.cs:22]: Here the same

File: BLL/Services/Log.cs (hunk 15 above)
Comment (stetskivostap) [reply BLL/Services/Log.cs:34]: Because Yesterday we decided to use ApendAllText

File: Common/Constants/FileConstants.cs [hunk 18]
```diff
@@ -0,0 +1,9 @@
+﻿namespace Common.Constants
+{
+    public class FileConstants
+    {
+        public const string FileNamePattern = "Exceptition";
```
Comment (vladpavlyuk) [inline Common/Constants/FileConstants.cs:5]:
```suggestion
        public const string FileNamePattern = "Exception_{0}.txt'
```

File: BLL/Services/Log.cs [hunk 19]
```diff
@@ -0,0 +1,37 @@
...
+    {
+        private readonly IWebHostEnvironment _web;
+        public Log(IWebHostEnvironment web)
+        {
+            _web = web;
+        }       
+        
+        private string GetFilePath()
+        {
+            var fileName = string.Format("{0}_{1}.txt", FileConstants.FileNamePattern, DateTime.Now.ToString(FileConstants.DateTimePattern));
```
Comment (vladpavlyuk) [inline BLL/Services/Log.cs:21]:
```suggestion
            var fileName = string.Format(FileConstants.FileNamePattern, DateTime.Now.ToString(FileConstants.DateTimePattern));
```

File: BLL/Services/Log.cs [hunk 20]
```diff
@@ -0,0 +1,37 @@
...
+            var logFilePath = string.Format(@"{0}{1}{2}", _web.ContentRootPath, FileConstants.LogFiles, fileName);
```
Comment (vladpavlyuk) [inline BLL/Services/Log.cs:22]: Use string concatenation

File: Common/Constants/ErrorConstants.cs [hunk 21]
```diff
@@ -0,0 +1,9 @@
+﻿namespace Common.Constants
+{
+    public static class ErrorConstants
+    {
+        public const string InternalServerError = "Oops something went wrong. Internal server error";
+
+        public const string Exception = "Exception_";
```
Comment (vladpavlyuk) [inline Common/Constants/ErrorConstants.cs:7]: Delete this constant

File: Common/Constants/FileConstants.cs (hunk 18 above)
Comment (vladpavlyuk) [inline Common/Constants/FileConstants.cs:5]: Rename it to ErrorLogFileNamePattern or something similiar

File: Common/Constants/FileConstants.cs [hunk 22]
```diff
@@ -0,0 +1,9 @@
+﻿namespace Common.Constants
+{
+    public class FileConstants
```
Comment (vladpavlyuk) [inline Common/Constants/FileConstants.cs:3]: Move Exception related constants ErrorConstants

File: BLL/Services/Log.cs [hunk 23]
```diff
@@ -0,0 +1,37 @@
...
+        private readonly IWebHostEnvironment _web;
+        public Log(IWebHostEnvironment web)
+        {
+            _web = web;
+        }       
+        
+        private string GetFilePath()
+        {
+            var fileName = string.Format(ErrorConstants.ErrorLogFileNamePattern, DateTime.Now.ToString(FileConstants.DateTimePattern));
+            var logFilePath = string.Format(_web.ContentRootPath + FileConstants.LogFiles + fileName);
```
Comment (vladpavlyuk) [inline BLL/Services/Log.cs:22]: Use concat instead of format method

File: Common/Constants/FileConstants.cs [hunk 24]
```diff
@@ -0,0 +1,8 @@
...
+    {        
+        public const string DateTimePattern = "dd_MM_yyyy";
+        public const string LogFiles = @"\LogFiles\";
```
Comment (vladpavlyuk) [inline Common/Constants/FileConstants.cs:6]: Move to error constants

File: BLL/Services/Log.cs [hunk 25]
```diff
@@ -18,8 +18,8 @@ public Log(IWebHostEnvironment web)
         
         private string GetFilePath()
         {
-            var fileName = string.Format(ErrorConstants.ErrorLogFileNamePattern, DateTime.Now.ToString(FileConstants.DateTimePattern));
-            var logFilePath = string.Format(_web.ContentRootPath + FileConstants.LogFiles + fileName);
+            var fileName = string.Concat(ErrorConstants.ErrorLogFileNamePattern, DateTime.Now.ToString(FileConstants.DateTimePattern));
+            var logFilePath = string.Concat(_web.ContentRootPath + ErrorConstants.LogFiles + fileName);
```
Comment (vladpavlyuk) [inline BLL/Services/Log.cs:22]: Why do you make a concatenation inside of Concat method? Check it https://docs.microsoft.com/ru-ru/dotnet/api/system.string.concat?view=netcore-3.1#System_String_Concat_System_String_System_String_System_String_System_String_
